runtime/bundles/org.eclipse.e4.core.di/src/org/eclipse/e4/core/internal/di/InjectorImpl.java

```diff
@@ -1026,4 +1019,31 @@ private boolean isAnnotationPresent(AnnotatedElement annotatedElement,
 		cache.put(annotatedElement, isPresent);
 		return isPresent;
 	}
+
+	/**
+	 * The IdentityWeakReference extends the {@link WeakReference} with the
+	 * difference that it can be compared to another IdentityWeakReference. The
+	 * compare is done on the Identity of the Referenced Object, this allows us to
+	 * use this element in a Set and have unique objects in it.
+	 */
+	private static class IdentityWeakReference<T> extends WeakReference<T> {
+
+		IdentityWeakReference(T referent) {
+			super(referent);
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (obj instanceof IdentityWeakReference other) {
+				return Objects.equals(other.get(), get());
+			}
+			return false;
+		}
+
+		@Override
+		public int hashCode() {
+			return System.identityHashCode(get());
```

Comment:
The problem here is that if (and of course when) the weak reference is garbage collected, get() will return null and then this class will returning different hashCodes at different times.  That will corrupt the set.   I think it will be necessary to compute the hash code and cache it in the constructor so that it can never change.

Reply:
Good idea.